core/src/com/unciv/logic/map/MapGenerator.kt

```diff
@@ -371,15 +372,66 @@ class MapGenerator(val ruleset: Ruleset) {
-        val candidateTiles = tileMap.values.filter { it.resource == null && it.isLand && !it.getLastTerrain().impassable }
-        val totalNumberOfResources = candidateTiles.size * tileMap.mapParameters.resourceRichness
-        val resourcesPerType = (totalNumberOfResources/strategicResources.size).toInt()
+        // Extensive but fast preparations - not many terrains therefore all collections are very small
+        // What, including terrain features, can count as land or water, respectively?
+        // This would make little sense with the vanilla ruleset but with this we allow some pretty free modding
+        val baseLand = ruleset.terrains.values.asSequence().filter { it.type == TerrainType.Land }
+        val landFeatures = ruleset.terrains.values.asSequence().filter { it.type == TerrainType.TerrainFeature && it.occursOn != null && it.occursOn.any { it in baseLand.map { it.name } } }
+        val allLand = (baseLand+landFeatures).map { it.name }.toHashSet()
+        val baseWater = ruleset.terrains.values.asSequence().filter { it.type == TerrainType.Water }
+        val waterFeatures = ruleset.terrains.values.asSequence().filter { it.type == TerrainType.TerrainFeature && it.occursOn != null && it.occursOn.any { it in baseWater.map { it.name } } }
+        val allWater = (baseWater+waterFeatures).map { it.name }.toHashSet()
+
+        // Now, per resource: Can they occur on land and/or water?
+        data class LandOrWater(val land: Boolean = false, val water: Boolean = false)
+        fun getLandOrWater(resource: TileResource): LandOrWater {
+            return LandOrWater(resource.terrainsCanBeFoundOn.any { it in allLand }, resource.terrainsCanBeFoundOn.any { it in allWater} )
+        }
+        val strategicResources =
+                ruleset.tileResources.values
+                .filter { it.resourceType == ResourceType.Strategic }
+        val landOrWaterMap =
+                strategicResources
+                .map { Pair<String,LandOrWater>(it.name, getLandOrWater(it)) }
+                .toMap()
+
+        // Get all base terrain any strategic resource can occur on / all features
+        // (could use only allOccursOn, benefit is small, but every little bit counts to speed up filtering later)
+        val allOccursOn = strategicResources.flatMap { it.terrainsCanBeFoundOn }.toHashSet()
+        val allBaseOccursOn = allOccursOn.map { ruleset.terrains[it]!! }.filter { it.type == TerrainType.Water || it.type == TerrainType.Land }.map { it.name }.toHashSet()
+        val allFeaturesOccursOn = allOccursOn.map { ruleset.terrains[it]!! }.filter { it.type == TerrainType.TerrainFeature }.map { it.name }.toHashSet()
+
+        // Get filtered sequences for allowable tiles without resources yet (nat wonders aren't spawned yet) - for all domain types
+        // Could be simplified by basing landCandidates/waterCandidates on allCandidates, but it should enumerate faster with cheapest tests first
+        val allCandidates =
```

Comment:
Since you do basically the same check each time with minor variations, this could be simplified with
`fun isCandidate(tile:Tile) = tile.resource==null && tile.baseTerrain in allBaseOccursOn || it.terrainFeature in allFeaturesOccursOn`

Reply:
Would work, and looked simpler earlier. Just optimized as much as possible - and one hash lookup is cheaper than two.